tests/exchanges/Ethfinex.test.js

```diff
@@ -0,0 +1,234 @@
+const Ethfinex = require('../../exchanges/Ethfinex.js')
+const { ETHFINEX_URL } = require('../../constants.js')
+
+describe('Ethfinex', () => {
+  const exchange = new Ethfinex()
+
+  const symbol = 'abc'
+  const amount = 500
+  const fee = 0
+
+  const noPriceDataMessage = 'no price data found on Ethfinex for abc'
+  const noLiquidityMessage = 'not enough liquidity on Ethfinex for 500 abc'
+
+  test('defines the proper markets endpoint', () => {
+    expect(exchange.marketsUrl).toBe(`${ETHFINEX_URL}/symbols`)
+  })
+
+  test('defines the proper orderbook endpoint', () => {
+    expect(exchange.orderbookUrl).toBe(`${ETHFINEX_URL}/book`)
+  })
+
+  describe('computePrice', () => {
+    const markets = ['abceth', 'defeth']
+    const bids = [{ price: '1.00', amount: '300' }, { price: '2.00', amount: '200' }]
+    const asks = [{ price: '3.00', amount: '200' }, { price: '4.00', amount: '300' }]
+    const orderBook = { asks, bids }
+
+    describe('errors with no price data', () => {
+      test('when an error occurs while getting the available markets', () => {
+        const getAvailableMarkets = jest.spyOn(exchange, 'getAvailableMarkets')
+        getAvailableMarkets.mockRejectedValue(new Error('something went wrong in getAvailableMarkets'))
+
+        expect(exchange.computePrice(symbol, amount, false, fee)).resolves.toMatchObject({
+          Ethfinex: {
+            exchangeName: 'Ethfinex',
+            tokenAmount: amount,
+            tokenSymbol: symbol,
+            timestamp: expect.any(Number),
+            error: noPriceDataMessage,
+          },
+        })
+      })
+
+      test('when an ETH market for the token does not exist', () => {
+        const getAvailableMarkets = jest.spyOn(exchange, 'getAvailableMarkets')
+        const badMarkets = ['defeth', 'xyzeth', 'abcbtc', 'defbtc', 'xyzbtc']
+        getAvailableMarkets.mockResolvedValue(badMarkets)
+
+        // TODO: Optimally we can spy on the error being thrown to assert the code flow
```

Comment:
@grahamperich  I was having trouble testing this with Jest in this method and the one below. Any chance you could help explain how I would go about doing this? I tried spying like this and could see the function being entered in the debugger, but when I tried to do `expect(_validateMarketAvailability).toHaveBeenCalled())` I was getting an error that the mock had never been interacted with. :/

Reply:
I'm not totally sure to be honest. If we were building this with tests from the first commit we probably would have structured things differently to accommodate every possible case. As it stands, this PR adds good coverage and I'm glad to merge it in